spec/compliance_data_spec.rb

```diff
@@ -0,0 +1,162 @@
+require 'spec_helper'
+require 'stub_classes'
+
+describe ComplianceData do
+  describe 'initialize' do
+    it 'returns the right thing' do
+      data = ComplianceData.new StubClasses::StubSettings
+      expect(data).to be_an_instance_of ComplianceData
+    end
+
+    it 'raises an argument error if settings are not correct' do
+      expect { ComplianceData.new nil }.to raise_error 'ArgumentError'
+    end
+  end
+
+  describe 'base_name' do
+    @data = nil
+
+    before :each do
+      @data = ComplianceData.new StubClasses::StubSettings
+    end
+
+    it 'does the right thing' do
+      base = @data.base_name 'path/to/file.json'
+      expect(base).to eq 'file'
+    end
+
+    it 'handles a missing extension' do
+      base = @data.base_name 'path/to/file'
+      expect(base).to eq 'file'
+    end
+
+    it 'handles a base name' do
+      base = @data.base_name 'file'
+      expect(base).to eq 'file'
+    end
+
+    it 'returns empty string if passed nil' do
+      base = @data.base_name nil
+      expect(base).to eq ''
+    end
+  end
+
+  describe 'full_name' do
+    @data = nil
+
+    before :each do
+      @data = ComplianceData.new StubClasses::StubSettings
+    end
+
+    it 'does the right thing' do
+      full = @data.full_name 'file'
+      expect(full).to eq 'results/file.json'
+    end
+
+    it 'returns an empty string if passed nil' do
+      full = @data.full_name nil
+      expect(full).to eq ''
+    end
+  end
+
+  describe 'key_list' do
+    @data = nil
+
+    class StubObjectSummary
+      attr_reader :key
+      def initialize(key_value)
+        @key = key_value
+      end
+    end
+
+    before :each do
+      @data = ComplianceData.new StubClasses::StubSettings
+    end
+
+    it 'does the right thing' do
+      allow_any_instance_of(Aws::S3::Bucket).to receive(:objects)
+        .and_return(
+          [
+            StubObjectSummary.new(
+              "#{StubClasses::StubSettings.results_folder}/abc"),
+            StubObjectSummary.new(
+              "#{StubClasses::StubSettings.results_folder}/bcd")
+          ])
```

Comment:
> I'm hesitant to change that model just to make the class easier to test

Ha, I actually think that is a _good_ reason to change code. From [the RSpec documentation for `allow_any_instance_of`](https://relishapp.com/rspec/rspec-mocks/docs/working-with-legacy-code/any-instance):

> This feature is sometimes useful when working with legacy code, though in general we
discourage its use

I think this is basically about dependency injection, which is especially useful when dealing with classes that make calls to external services.

> can't you just say `@data.bucket.should` receive

That works too!

We should also add [WebMock](https://github.com/bblimke/webmock) to ensure that no real network requests are made.